url147ce07ebe_JoshuaKissoon_Kademlia_tgz-pJ8-kademlia_simulations_RoutingTableStateTestingJ8/src/kademlia/operation/NodeLookupOperation.java

```diff
@@ -0,0 +1,323 @@
+    private static final String AWAITING = "Awaiting";
+    private static final String ASKED = "Asked";
+    private static final String FAILED = "Failed";
+
+    private final KadServer server;
+    private final KademliaNode localNode;
+    private final KadConfiguration config;
+
+    private final Message lookupMessage;        // Message sent to each peer
+    private final Map<Node, String> nodes;
+
+    /* Tracks messages in transit and awaiting reply */
+    private final Map<Integer, Node> messagesTransiting;
+
+    /* Used to sort nodes */
+    private final Comparator comparator;
+
+    
+    {
+        messagesTransiting = new HashMap<>();
+    }
+
+    /**
+     * @param server    KadServer used for communication
+     * @param localNode The local node making the communication
+     * @param lookupId  The ID for which to find nodes close to
+     * @param config
+     */
+    public NodeLookupOperation(KadServer server, KademliaNode localNode, KademliaId lookupId, KadConfiguration config)
+    {
+        this.server = server;
+        this.localNode = localNode;
+        this.config = config;
+
+        this.lookupMessage = new NodeLookupMessage(localNode.getNode(), lookupId);
+
+        /**
+         * We initialize a TreeMap to store nodes.
+         * This map will be sorted by which nodes are closest to the lookupId
+         */
+        this.comparator = new KeyComparator(lookupId);
+        this.nodes = new TreeMap(this.comparator);
+    }
+
+    /**
+     * @throws java.io.IOException
+     * @throws kademlia.exceptions.RoutingException
+     */
+    @Override
+    public synchronized void execute() throws IOException, RoutingException
+    {
+        try
+        {
+            /* Set the local node as already asked */
+            nodes.put(this.localNode.getNode(), ASKED);
+
+            /**
+             * We add all nodes here instead of the K-Closest because there may be the case that the K-Closest are offline
+             * - The operation takes care of looking at the K-Closest.
+             */
+            this.addNodes(this.localNode.getRoutingTable().getAllNodes());
+
+            /* If we haven't finished as yet, wait for a maximum of config.operationTimeout() time */
+            int totalTimeWaited = 0;
+            int timeInterval = 10;     // We re-check every n milliseconds
+            while (totalTimeWaited < this.config.operationTimeout())
+            {
+                if (!this.askNodesorFinish())
+                {
+                    wait(timeInterval);
+                    totalTimeWaited += timeInterval;
+                }
+                else
+                {
+                    break;
+                }
+            }
+
+            /* Now after we've finished, we would have an idea of offline nodes, lets update our routing table */
+            this.localNode.getRoutingTable().setUnresponsiveContacts(this.getFailedNodes());
+
+        }
+        catch (InterruptedException e)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)